design/sync-v2.md

```diff
@@ -0,0 +1,216 @@
+* A `Deleted` flag may be provided by using `[Deleted]` on a field.
+
+For a `TodoItem`, for example:
+
+```csharp
+class TodoItemDTO : IDatasyncVersion
+{
+    [Key]
+    public long Id { get; set; }
+
+    [UpdatedAt]
+    public DateTimeOffset UpdatedAt { get; set; }
+
+    public string IDatasyncVersion.GetVersion()
+        => UpdatedAt.UtcTicks.ToString();
+
+    public string Title { get; set; }
+
+    public string Description { get; set; }
+
+    public DateOnly DueDate { get; set; }
+
+    public TimeSpan AlertBeforeDueDate { get; set; }
+
+    public TimeOnly AlertTime { get; set; }
+
+    public DateTimeOffset? CompletedDate { get; set; }
+
+    public UIColor StatusColor { get; set; }
+}
+```
+
+You will notice:
+
+* We rely on the database to handle the ID and UpdatedAt.
+* The version is computed from the UpdatedAt field - no need to keep something extra.
+* There are a variety of types here; we should artificially restrict the types.
+* The DTO should be able to be handled on both client and server, potentially mapping using Automapper to get the DatabaseGenerated field for UpdatedAt.
+
+This should also be handled quite easily with non-OData requests; for example, GraphQL, gRPC, and other transports should be easily provided.
+
+## Local Service
+
+On the local side, we should utilize common libraries, like Entity Framework Core, to implement the offline synchronization library.  This is split into two parts - 
+
+1. When the client application accesses the local database
+2. When the synchronization process makes changes to the local database.
+
+Ideally, the same DTO should be able to be used on both sides (remote service and offline database).  
+
+When a client application makes a change to the local database:
+
+1. The change is validated.
+2. The operation is written to the operations queue.
+3. The change is written to the local offline table.
+
+When the synchronization process makes a change to the local database:
+
+1. The change is validated.
+2. The operation is removed from the operations queue.
+3. The change is written to the local offline table.
+
+In both cases, the operations queue and offline table change needs to be written in a transaction so that it may be rolled back in its entirety. 
+
+In the new world:
+
+### Pull Operation
+
+1. Read delta-token from the offline table.
+2. Read new items since delta-token from remote service.
+3. Add new items to the offline table.
+4. Write delta-token to the offline table.
+
+### CRUD on offline table
+
+1. Add new entity to the offline table, writing an operation table entry as well (as transaction).
+    TODO: How to deal with Id field that is not globally unique in the offlien table.
+
+### Push Operation
+
+1. Read (block of) entities in the operations queue
+1. For each entity:
+    * Do operation on remote service.
+    * On success, store new entity + remote operation table entry as a transaction.
+    * On conflict, deal with conflicts as per current spec.
```

Comment:
An automatic conflict resolution for one special case could be discussed:
We have often unnecessary conflict resolutions due to maintenance jobs on the SQL-server backend. The maintenance jobs lead to an update of the version field but change no content in the subset of fields that are pulled by clients. If on the client side a non-content-altering save to the offline store happens (User press OK instead of Cancel), `Version` is updated anyway. 
So, a conflict arises, that could be resolved automatically as a result of an object deep compare (for payload only): `If Equals(..) == true { localItem.Version = serverItem.Version; store localItem; Handled = true; }`. 
A datasync client specific object deep compare could be a decision support in case of conflicts.